### ISSUE TITLE
kebab-case option warns of single words in component-selector

### ISSUE THREAD

**Issue by gSarciotto** [angular-eslint/angular-eslint#318]
While creating the docs for the rule [component-selector](https://github.com/angular-eslint/angular-eslint/blob/master/packages/eslint-plugin/src/rules/component-selector.ts), I found an error message different from what I expected

#### Expected message

`The selector should be prefixed by one of the prefixes: 'app,toh' (https://angular.io/guide/styleguide#style-02-07) eslint@angular-eslint/component-selector`

#### Received message

`The selector should be kebab-case (https://angular.io/guide/styleguide#style-05-02) eslint@angular-eslint/component-selector`

#### Configuration
```json
"rules": {
    "@angular-eslint/component-selector": [
        "warn",
        {
            "type": "element",
             "style": "kebab-case",
             "prefix": ["app", "toh"]
        }
    ]
}
```

#### Code
```ts
@Component({
  selector: 'root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {}
```

Digging through the code, I found this to be a [deliberate decision](https://github.com/angular-eslint/angular-eslint/blob/c44cd5d043a3d203efd1faaed4cbfee2c9ac2e9d/packages/eslint-plugin/tests/rules/component-selector.test.ts#L285-L296). Why is it? I mean, a single word like `app` is a valid kebab-cased word (from what I known).

**Comment by rafaelss95** [comment]
Nice finding, @gSarciotto! I'm not sure why this decision was made, but IMHO we have an inconsistent behavior here.

The [`SelectorValidator.kebabCaseValidator`](https://github.com/angular-eslint/angular-eslint/blob/c44cd5d043a3d203efd1faaed4cbfee2c9ac2e9d/packages/eslint-plugin/src/utils/utils.ts#L546) requires an alphanumeric/hyphen after the _real_ hyphen.

Tests for `kebabCase`:

- 00-0 ✅  
- app-- ✅  
- app-0 ✅  
- appA ❌ 
- app0 ❌ 
- app ❌ 

On the other hand, the [`SelectorValidator.camelCase`](https://github.com/angular-eslint/angular-eslint/blob/c44cd5d043a3d203efd1faaed4cbfee2c9ac2e9d/packages/eslint-plugin/src/utils/utils.ts#L538) doesn't care about having an uppercased letter:

Tests for `camelCase`:
- 00-0 ❌ 
- app-- ❌ 
- app-0 ❌ 
- appA ✅ 
- app0 ✅ 
- app ✅

**Comment by gSarciotto** [comment]
@rafaelss95 Yeah, this difference in behavior is inconsistent. 

IMHO `camelCase` shouldn't enforce having at least one uppercased letter and `kebab-case` shouldn't enforce having at least one dash

**Comment by JamesHenry** [comment]
Sounds good to me @gSarciotto please feel free to submit a PR

**Comment by JamesHenry** [comment]
I want to do a v3 release before we switch to v12 (to line up with Angular), so out of an abundance of caution let's class this as a breaking change. But as I say feel free to submit a PR whenever you can

**Comment by gSarciotto** [comment]
> Sounds good to me @gSarciotto please feel free to submit a PR

Alright. I'm busy but I think I can manage to do it this week

**Comment by gSarciotto** [comment]
@JamesHenry @rafaelss95 I think I discovered why kebab case enforces the use of a dash, it probably is to avoid conflicts with native HTML selectors like `label`. This is actually mentioned in the [styleguide](https://angular.io/guide/styleguide#style-02-07).

So should I remove this restriction to the `kebab-case` or enforce it in the `camelCase` options?

**Comment by JamesHenry** [comment]
Thanks for digging deeper @gSarciotto - it seems to me then that actually it's just the lint message that needs improving.

As you've said having just `root` goes against the intention of the rule so feedback needs to be given. In subsequent comments we have been focusing too strictly on the definition of the casing terms rather than looking at what the rule is trying to enforce.

The message should be something more along the lines of:

`The selector should be kebab-case and be prefixed by one of the following: 'app,toh'`

We can class a change to the message as a non-breaking enhancement so feel free to submit a PR whenever you have time.

**Comment by rafaelss95** [comment]
Nice finding, thanks, @gSarciotto!

> it probably is to avoid conflicts with native HTML selectors like `label`. This is actually mentioned in the [styleguide]

@JamesHenry In that case, shouldn't we force camelCase to have a real prefix too? If not one will still be able to type conflicting selectors like `label`, `input` and so on, something that we are avoiding with the `kebab-case` style.

Also I just noticed that we perform some suffix validations on the `prefix` function:

https://github.com/angular-eslint/angular-eslint/blob/c44cd5d043a3d203efd1faaed4cbfee2c9ac2e9d/packages/eslint-plugin/src/utils/utils.ts#L563-L567

**Comment by sam-s4s** [comment]
So, I've just run into this problem... I an not using prefixes, and thus my rule is...
```
"@angular-eslint/component-selector": [
                    "error",
                    {
                        "prefix": [ "" ],
                        "style": "kebab-case",
                        "type": "element"
                    }
                ],
```

So firstly, this rule should not force you to use prefixes, as that is a choice - and I've found that the only way I can circumvent this is to provide an array with an empty string. Anything else actually just silently prevents this rule from running at all (ie. the kebab-case is not detected in "MyComponent" (which should be "my-component") - it's just ignored because the rule is not run if there are no prefixes defined.

And secondly - more relating to this issue - a single word should be considered valid kebab case, so long as it has no uppercase letters. ie "blah" is good, and "Blah" is not. But currently "blah" is giving me an error saying it's not kebab case - which is definitionally incorrect, as a single word definitely is kebab case.

The code in question is...
```
kebabCase(selector: string): boolean {
    return /^[a-z0-9-]+-[a-z0-9-]+$/.test(selector);
},
```
https://github.com/angular-eslint/angular-eslint/blob/master/packages/eslint-plugin/src/utils/utils.ts#L721

Sadly for now I'm going to have to disable this rule until this is fixed :(

**Comment by agataHappy** [comment]
Is there any update on that issue? I'm currently trying to add ESlint to repo and have this problem.

**Comment by zarkosusnjar** [comment]
Same issue here. 

Migrated from TSLint. Recommended settings complain about kebab-case in component selector 'footer'. 
Also, IMHO prefixes should be optional since we're dealing with styling developed by third party.

**Comment by sam-s4s** [comment]
Thanks to @abaran30 for fixing this! :)

Can I ask (while this is fresh in your head) - I think this problem also exists for @angular-eslint/directive-selector - would it be possible to do the same thing there?

**Comment by abaran30** [comment]
My pleasure, @sam-s4s!

This doesn't appear to be an issue for directive-selector. I was able to try the following test as a valid case, and it passed successfully:
```
  {
    code: `
      @Directive({
        selector: 'singleword'
      })
      class Test {}
      `,
    options: [{ type: 'element', style: 'kebab-case' }],
  },
```
Should you be able to reproduce this in any fashion for directive-selector, please create a new issue 🙂.

**Comment by sam-s4s** [comment]
Oh, my bad then! Maybe it's been fixed since I tried it last year :D

I shall check it out again just in case.

Thank you! <3
